src/std/crypto/pkey.ss

```diff
@@ -21,10 +26,116 @@
     (EVP_PKEY_keygen ctx)
     (foreign-release! ctx)))
 
-(def (bytes->private-key type bytes engine: (engine #f))
+;;; * Meta: pkey-[type|type-id|is-a?|?]
+
+(def pkey-type EVP_PKEY_get0_type_name)
+(def pkey-type-id EVP_PKEY_get_base_id)
+(def pkey-is-a? EVP_PKEY_is_a)
+(def (pkey? k) (and (EVP_PKEY? k) #t))
+
+;;; * EVP_PKEY_* Aliases: 'FOO = EVP_PKEY_FOO
+
+(def evp-pkey-alias-table (make-hash-table))
+
+(def (evp-pkey-alias name)
+  (def (find-alias)
+    (ignore-errors
+     (eval (string->symbol (format "std/crypto/libcrypto#EVP_PKEY_~a" name)))))
+  
+  (or (hash-get evp-pkey-alias-table name)
+      (let (t (find-alias))
+	(if (not t) t 
+	    (begin (hash-put! evp-pkey-alias-table name t)
+		   t)))))
+
+(def (evp-pkey-enum n)
+  (if (number? n) n
+      (evp-pkey-alias n)))
+
+;;; * EVP_* Aliases: 'sha256 = EVP_sha256
+
+(def evp-alias-table (make-hash-table))
+
+(def (evp-alias name)
+  (def (find-alias)
+    (ignore-errors
+     (eval (string->symbol (format "std/crypto/libcrypto#EVP_~a" name)))))
+  
+  (or (hash-get evp-alias-table name)
+      (let (t (find-alias))
+	(if (not t) t 
+	    (begin (hash-put! evp-alias-table name t)
+		   t)))))
+
+(def (evp-md md)
+  (and md (if (EVP_MD? md) md
+	      (or (let (proc (evp-alias md))
+		    (and proc (proc)))
+		  (error "Cannot find cryptographic hash model function" md)))))
+
+;;; * pkey types and subtypes
+
+(def pkey-types-with-raw-private-key
+  (map evp-pkey-enum
+       '(HMAC POLY1305 SIPHASH X25519 ED25519 X448 ED448)))
+
+(def (pkey-type-raw-private-key? type) 
+  (memv (evp-pkey-enum type) pkey-types-with-raw-private-key))
+
+(def pkey-types-with-raw-public-key
+  (map evp-pkey-enum
+       '(X25519 ED25519 X448 ED448)))
+
+(def (pkey-type-raw-public-key? type) 
+  (memv (evp-pkey-enum type) pkey-types-with-raw-public-key))
+
+;;; * bytes->[pkey|private-key|public-key]
+
+(def (bytes->pkey bytes
+		  type: (type #f)
+		  input-type: (input-type #f)
+		  raw?: (raw? #f)
+		  engine: (engine #f)
+		  selection: (selection 0))
+  (if raw?
+    ((cond ((eq? raw? 'private)
+	    bytes->private-key)
+	   ((eq? raw? 'public)
+	    bytes->private-key))
+     type bytes raw?: #t engine: engine)
+    (let (in (BIO_new_mem_buf bytes))
+      (EVP_PKEY_new_from_bio
+       in input-type type (evp-pkey-enum selection)))))
+
+(def (%bytes->pkey type bytes engine raw? input-type selection raw-fn)
```

Comment:
I prefer to use `_`  as prefix for "internal" thingies. Let's try to be consistent?